# The methodology behind `signet` 

## Biological networks conversion to graphs

In order to get information about relationships between genes,
we advise to use the Bioconductor package `graphite`, which implements a 
procedure to convert biological pathways data from different databases to 
graphs.

These databases have emerged as references in systems biology:

* KEGG database is a prominent reference knowledge base of large-scale 
molecular data sets generated by genome sequencing and other 
experimental technologies. KEGG Pathways provides maps for both
signaling and metabolic pathways.

* Reactome, backed by the EBI, is one of the most complete repository; 
it is frequently updated and provides a semantically rich description 
of each pathway.

* BioCarta is a developer, supplier and distributor company of reagents 
and assays for biopharmaceutical and academic research.

* NCI (NCI/Nature Pathway Interaction Database) is a highly-structured, 
curated collection of information about known 
biomolecular interactions and key cellular processes assembled into signaling 
pathways.

* Panther data are a comprehensive, curated database of pathways backed by the
University of Southern California.

* HumanCyc is part of the BioCyc database collection of pathways.

To transform pathway object to graphs, we use the procedure implemented in the
package graphite (for more details, see X et al.).

Within a pathway, nodes often correspond to multiple gene products. 
These can be divided into protein complexes (proteins linked by protein-protein 
interactions) and the groups containing alternative members (genes generally 
with similar biochemical functions). Thus, when considering signal propagation 
these groups should be considered differently; the first (hereafter group AND) 
should be expanded into a clique (all proteins connected to the others), while 
the second (hereafter group OR) should be expanded without connection among the 
contained elements (X et al, XXX).

Compound mediated interactions are interactions for which a compound acts as a 
bridge between two elements. Since we are interested in gene networks, they 
should be removed from the graph. However, the trivial elimination of the 
compounds, without signal propagation, will strongly bias the topology 
interrupting the signals that pass through them. If element A is linked to 
compound c and compound c is linked to element B, thus elements A should be 
linked to elements B.

X et al, (XXX) handle this problem as follows: since signal propagation 
reconstruction is crucial for topological gene set analysis we decide to include
additional rules for the propagation reconstructing a connection between two 
genes connected through a series of chemical compounds. Not all compound are 
considered for the propagation because some of them, such as Hydrogen, H2O, ATP,
ADP etc., are highly frequent in map descriptions and the signal propagation 
through them would lead to degenerate too long chains of compounds. The 
compounds not considered for propagation are not characteristic of a specific 
reaction but secondary substrates/products widely shared among different 
processes.

## Correction for overlapping between pathways

We will first exclude all pathways completely included in at least another 
pathway (overlap = 1). I still need to implement a procedure to merge pathways 
with an overlapping level greater than a given threshold.

Finally, pathways with no connected component of size greater than 10 
(arbitratry threshold) are removed.

## Background distribution

To apply the algorithm to a list of pathways, you have to first generate the 
background distribution of your summary statistic.

The input is the list of all the pathways you want to analyze, and a list of 
scores for each gene. The background distribution or the statistic will be 
generated for each possible subnetwork size (kmin to kmax).

A pathway is randomly sampled (the sampling probability being conditioned by 
the number of genes in the pathway) and a connected subnetwork of size k is 
randomly picked (one gene is sampled in the pathway, then k genes in the 
boundary). The score of the subnetwork is then computed. This is done N times 
for each k, to get the background distribution of subnetworks scores, we just 
keep the mean and standard error of this distribution.

## High-scoring subnetwork search

We consider that genes can yield two states: active or inactive.

The selected gene is not an articulation point of the subgraph, i.e. its 
removal doesnâ€™t disconnect the active subgraph.

The selected gene is randomly picked in the following nodes: i) nodes in 
the boundary; ii) leaves, iii) nodes which are not articulation points of 
the subgraph.

Figure

## Output interpretation

The output is, for each pathway tested, the subnetwork, its score and its 
significance.
Careful: Not the whole pathway is tested, just the subnetwork identified. 
Need to characterize the function of this subnetwork, which is not necessarily 
limited to the pathway function. For example, how can we interpret a subnetwork 
of 10 genes inside a 1000 genes pathway?

In the case of FST scan, this methodology is just a first approach to 
detect candidate subnetworks. Further analysis should include linkage 
disequilibrium characterization to confirm that the differentiation of 
a set of genes is the same between populations or not.

## Perspectives

Procedure to quantify the function of the subnetwork.
Edge scores: linkage disequilibrium between genes?
