# The methodology behind `signet` 

## Introduction

Pritchard & Di Rienzo (2010) argued that many, or most, adaptive events in natural populations occur by the evolution of polygenic traits, rather than via the fixation of single beneficial mutations . Recent genome-wide association studies in humans (Yang et al. 2010; Stranger et al. 2011) and in various model organisms (e.g. in Arabidopsis thaliana: Atwell et al. 2010; in Drosophila melanogaster: Jumbo-Lucioni et al. 2010) have indeed  confirmed  that  variation  at  many  important traits is controlled by a large number of loci dispersed throughout the genome. Polygenic adaptation typically involves small allele frequency changes at many loci (Mackay et al. 2009). Therefore, quantitative trait loci (QTL) involved in polygenic adaptation will go undetected using methods for detecting the molecular signature of selective sweeps at individual loci (Pritchard & Di Rienzo 2010).

Genome scans for selection usually aim at identifying single outlier loci. But as some phenotypic traits involve many loci, adaptive events can result in small changes at several loci (polygenic adaptation), leading to difficulties in detecting this type of adaptation. 

To detect this subtle changes in allele frequencies, one could

List of pathways. First, we need to correct for pathway overlapping. Databases. After, algorithm. s

## Pathway databases

In order to gather curated information about human pathways, we have collected data from the four public databases that have emerged as reference points for the systems biology community. The KEGG database is a prominent reference knowledge base for integration and interpretation of large-scale molecular data sets generated by genome sequencing and other high-throughput experimental technologies. KEGG Pathways provides maps for both signaling and metabolic pathways [1]. Reactome, backed by the EBI, is one of the most complete repository; it is frequently updated and provides a semantically rich description of each pathway [2]. BioCarta (www.biocarta.com) is a developer, supplier and distributor company of reagents and assays for biopharmaceutical and academic research. Through an "open source" approach, this community-fed forum constantly integrates emerging proteomic information from the scientific community. It also catalogs and summarizes important resources providing information for over 120,000 genes. BioCarta pathway data in biopax format are available through NCI website. NCI (NCI/Nature Pathway Interaction Database [3]) is a highly-structured, curated collection of information about known biomolecular interactions and key cellular processes assembled into signaling pathways. This was a collaborative project between the NCI and Nature Publishing Group (NPG). Finally, Panther [11] data are a comprehensive, curated database of pathways, protein families, trees, subfamilies and functions available at http://pantherdb.org backed by the University of Southern California. HumanCyc is part of the BioCyc database collection of pathways [10].

## Pathway data preparation

### Transformation to graph objects

To transform pathway object to graphs, we use the procedure implemented in the R package graphite (X et al, XXX).

Within a pathway, nodes often correspond to multiple gene products. These can be divided into protein complexes (proteins linked by protein-protein interactions) and the groups containing alternative members (genes generally with similar biochemical functions). Thus, when considering signal propagation these groups should be considered differently; the first (hereafter group AND) should be expanded into a clique (all proteins connected to the others), while the second (hereafter group OR) should be expanded without connection among the contained elements (X et al, XXX).

Compound mediated interactions are interactions for which a compound acts as a bridge between two elements. Since we are interested in gene networks, they should be removed from the graph. However, the trivial elimination of the compounds, without signal propagation, will strongly bias the topology interrupting the signals that pass through them. If element A is linked to compound c and compound c is linked to element B, thus elements A should be linked to elements B.

X et al, (XXX) handle this problem as follows: since signal propagation reconstruction is crucial for topological gene set analysis we decide to include additional rules for the propagation reconstructing a connection between two genes connected through a series of chemical compounds. Not all compound are considered for the propagation because some of them, such as Hydrogen, H2O, ATP, ADP etc., are highly frequent in map descriptions and the signal propagation through them would lead to degenerate too long chains of compounds. The compounds not considered for propagation are not characteristic of a specific reaction but secondary substrates/products widely shared among different processes.

### Overlapping correction

We will first exclude all pathways completely included in at least another pathway (overlap = 1). 
I still need to implement a procedure to merge pathways with an overlapping level greater than a given threshold. Or use a pruning method afterwards.

Finally, pathways with no connected component of size greater than 10 (arbitratry threshold) are removed.

## Background distribution

To apply the algorithm to a list of pathways, you have to first generate the background distribution of your summary statistic.

The input is the list of all the pathways you want to analyze, and a list of scores for each gene. The background distribution or the statistic will be generated for each possible subnetwork size (kmin to kmax).

A pathway is randomly sampled (the sampling probability being conditioned by the number of genes in the pathway) and a connected subnetwork of size k is randomly picked (one gene is sampled in the pathway, then k genes in the boundary). The score of the subnetwork is then computed. This is done N times for each k, to get the background distribution of subnetworks scores, we just keep the mean and standard error of this distribution.

## High-scoring subnetwork search

We consider that genes can yield two states: active or inactive.

The selected gene is not an articulation point of the subgraph, i.e. its removal doesnâ€™t disconnect the active subgraph.

The selected gene is randomly picked in the following nodes: i) nodes in the boundary; ii) leaves, iii) nodes which are not articulation points of the subgraph.

Figure

## Testing the significance of the subnetwork score

As the high-scoring search procedure tends to bias the p-values distribution towards low values, the test will be non-conservative if we use the background distribution to compute the significance of the subnetwork score. Therefore, the test implemented uses this high-scoring search procedure. For N iterations, gene scores are permuted. Then, a pathway is randomly sampled (the sampling probability being conditioned by the number of genes in the pathways) and the search algorithm is applied to this pathway. The score of the high-scoring subnetwork found in the randomized data is computed and its distribution is generated. This is the null distribution of the test.

Finally, a correction for multiple testing is considered as we applied the test to hundreds of pathways. We use the q-value with the method of X et al. (XXX) implemented in the R package qvalue. We consider a FDR of 0.001 to be even more conservative.

## Output interpretation

The output is, for each pathway tested, the subnetwork, its score and its significance.
Careful: Not the whole pathway is tested, just the subnetwork identified. Need to characterize the function of this subnetwork, which is not necessarily limited to the pathway function. For example, how can we interpret a subnetwork of 10 genes inside a 1000 genes pathway?

In the case of FST scan, this methodology is just a first approach to detect candidate subnetworks. Further analysis should include linkage disequilibrium characterization to confirm that the differentiation of a set of genes is the same between populations or not.

## Perspectives

Procedure to quantify the function of the subnetwork.
Edge scores: linkage disequilibrium between genes?
